google-cloud-storage/src/main/java/com/google/cloud/storage/testing/FakeStorageRpc.java

```diff
@@ -0,0 +1,658 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage.testing;
+
+import com.google.api.services.storage.model.Bucket;
+import com.google.api.services.storage.model.BucketAccessControl;
+import com.google.api.services.storage.model.HmacKey;
+import com.google.api.services.storage.model.HmacKeyMetadata;
+import com.google.api.services.storage.model.Notification;
+import com.google.api.services.storage.model.ObjectAccessControl;
+import com.google.api.services.storage.model.Policy;
+import com.google.api.services.storage.model.ServiceAccount;
+import com.google.api.services.storage.model.StorageObject;
+import com.google.api.services.storage.model.TestIamPermissionsResponse;
+import com.google.cloud.Tuple;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.spi.v1.RpcBatch;
+import com.google.cloud.storage.spi.v1.StorageRpc;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.math.BigInteger;
+import java.nio.file.FileAlreadyExistsException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A bare-bones in-memory implementation of StorageRpc, meant for testing.
+ *
+ * <p>This class is <i>not</i> thread-safe. It's also (currently) limited in the following ways:
+ *
+ * <ul>
+ *   <li>Supported
+ *       <ul>
+ *         <li>object create
+ *         <li>object get
+ *         <li>object delete
+ *         <li>list the contents of a bucket
+ *         <li>generations
+ *       </ul>
+ *   <li>Unsupported
+ *       <ul>
+ *         <li>bucket create
+ *         <li>bucket get
+ *         <li>bucket delete
+ *         <li>list all buckets
+ *         <li>file attributes
+ *         <li>patch
+ *         <li>continueRewrite
+ *         <li>createBatch
+ *         <li>checksums, etags
+ *         <li>IAM operations
+ *         <li>BucketLock operations
+ *         <li>HMAC key operations
+ *       </ul>
+ * </ul>
+ */
+public class FakeStorageRpc implements StorageRpc {
+
+  // fullname -> metadata
+  Map<String, StorageObject> metadata = new HashMap<>();
+  // fullname -> contents
+  Map<String, byte[]> contents = new HashMap<>();
+  // fullname -> future contents that will be visible on close.
+  Map<String, byte[]> futureContents = new HashMap<>();
+
+  private final boolean throwIfOption;
```

Comment:
not sure what this field name means

Reply:
FakeStorageRpc.java is copied almost as-is from: 
https://github.com/googleapis/java-storage-nio/blob/master/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/testing/FakeStorageRpc.java